cmd/ingester/app/consumer/rate_limiter.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2018 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consumer
+
+import (
+	"time"
+)
+
+// rateLimiter uses a ticker to keep events evenly spaced, thereby providing
+// rate limiting.
+type rateLimiter struct {
+	ticker      *time.Ticker
+	tickChannel <-chan time.Time
+	done        chan struct{}
+}
+
+// newRateLimiter constructs a rate limiter with the provided creditsPerSecond.
+// If creditsPerSecond is less than zero, no rate limiting will be
+// provided, and all calls to this object will return immediately. If
+// creditsPerSecond is exactly zero, calls will block indefinitely or until Stop
+// is called.
+func newRateLimiter(creditsPerSecond float64) *rateLimiter {
+	var r *rateLimiter
```

Comment:
I think we can simplify this a bit. Initialize empty tickChannel and ticker at the top and go through the elseif craziness. At the bottom, initialize the rateLimiter. Should reduce the amount of `done:        make(chan struct{})` everywhere. Also, I prefer switch statement here instead of if else (just personal preference).

Reply:
Sure. I forgot Go allowed switches with conditions instead of values.

Reply:
I still feel that this general approach is more complex than using a simple token bucket without involving channels. 
What benefits do we gain by doing this? 

See https://github.com/uber-go/ratelimit/blob/c15da02342779cb6dc027fc95ee2277787698f36/ratelimit.go for an example - IMHO it is more readable.

Reply:
I tend to agree. That rate limiter just sleeps until the next token is available. Is that worse than using channels for signaling?

Reply:
I will admit the rate limiter here is also less accurate. The ticker may drop ticks if we aren't receiving from the tick channel frequently enough. My issue with redoing the token bucket is that we should ideally just use the one in the client, and potentially migrate it to jaeger-lib. This is the second time I have reimplemented a token bucket in the project (throttling was first).

Reply:
There is a slightly different semantic, so I guess "less accurate" is not a good way of putting it. The token bucket allows for bursts if enough credits have accumulated. A ticker will not allow for any bursts.

Reply:
Oh man... I already did that in https://github.com/uber/jaeger-lib/pull/35.

Reply:
This is excellent - could we reuse that ratelimiter?

Reply:
If we weren't expecting a blocking rate limiter, I would have just reused that. The blocking involves calculating a interval and waiting for a timer. The existing rate limiter does not expose its credit balance, making this more difficult.

Reply:
Is it very difficult to extend the ratelimiter in `jaeger-lib` to do this?